feature: Add icon opacity and desaturation settings

Finii opened the pull request:
Sometimes old applications with their shiny tray icons do not blend very well into the modern, "materials-style" desktop of today. It would be useful to be able to de-saturate that icons and or adjust the contrast, brightness, and opacity.

That doesn't fix the design problem in those older apps but this greatly mitigates the visual annoyance.

Some people know that from the no longer maintained TopIcon-Plus extension.

This pull request adds a configuration dialog, where one can set
* Opacity
* Saturation
* Brightness
* Desaturation

as global settings for all icons. Here is a visual impression of icons and dialog:

![Screenshot from 2019-10-30 15-22-57](https://user-images.githubusercontent.com/16012374/67869960-77e3dc80-fb2e-11e9-9ef1-6397a592cde3.png)
_Two desaturated icons (nomachine, nextcloud) and the Gnome settings dialog_

The code and translations are taken from TopIcon-Plus. https://github.com/phocean/TopIcons-plus

Fixes #142

arigit commented:
Tested and in use on Ubuntu 19.10 with GNOME login. Works perfectly well!
Thanks @Finii 

@3v1n0 hoping this can get merged

goodwillcoding commented:
@Finii ... love the feature but would suggest that a white list should be added. This would help icons such slack where most of the icon fits the "material style but it also uses colored dot to indicated messages.

Finii commented:
Not sure what you meant. Dropping the autogeneratable files. Still changing the Makefile, hmm.

Finii commented:
Resolved conflicts. Force push.

Atrate commented:
@Finii Would it also be possible to get a setting for icon size?

Finii commented:
@Atrate: You mean the same "Icon Size (min: 0, max: 96)" that TopIcon Plus has? Or do you have another functionality in mind?

I reckon it's easy to add.
On the other hand I'm not sure adding more stuff increases the probability of this PR being actually pulled :->

Atrate commented:
Yes, that one, since I've seen appIndicators often being bigger than other shell icons.

Finii commented:
@Atrate: As easy as can be ;)
Looks good on my theme. Please check it out.

Also closes #205.

Also looked into #200, but the spacing is a bit more involved - and noone requested it here ;-D
My personal interest is rather even more slim for icon spacing.
Icon size on the other hand, although I never missed it, is indeed a nice addition.

3v1n0 commented:
I'm sorry, but latest changes create conflicts... So it would be nice if you could rebase on master.

cyfrost commented:
Totally off-topic but this PR reminds me if we can also have a feature request for customizing the spacing between the icons.

Finii commented:
Rebasing on master is not too hard, did it, but the change from ``Shell.Stack`` to ``St.Icon`` needs some more changes. So this might take some time.

Finii commented:
Rebase on master. The change from ``Shell.Stack`` to ``St.Icon`` need a little bit more change than just whitespaces. So this comes in two batches.

Original commits rebased (and not working!):
* Add desaturation (hard coded) 2ae7b4e
* Add settings dialog 8457f33
* make: Handle schema compilation 7935ff9
* make: Autogenerate translation files d535d91
* make: Remove autogenerated files on target clean ac620bd
*Add possibility to limit icon siz 8bb7566
* appIndicator: Remove unused code a4ba87e
* appIndicator: Fix instantaneous settings changes d950c83 

And these commits reestablish functioning code:
* appIndicator: Remove unused icon modification code paths 1f0e4fd
* appIndicator: Adapt saturation etc to St.Icon setup 409a054
* Revert "Add possibility to limit icon size" 1f3d4a2 

Note: The last commit removes the not working - in my opinion not anymore needed - possibility to change the icon max size. I gave it a try, but that needs either a better idea or some more code than I currently can put into this feature. Sorry :-(

terencode commented:
Your latest rebase is working well but I'm getting a gnome-shell crash when playing with the different values (clicking and holding + and -), can you reproduce ?

```
#0  0x00007f625a477ce5 in raise () at /usr/lib/libc.so.6
#1  0x0000560cbe6559e9 in  ()
#2  0x00007f625a477d70 in <signal handler called> () at /usr/lib/libc.so.6
#3  0x00007f6259f02f46 in  () at /usr/lib/libglib-2.0.so.0
#4  0x00007f6259f03136 in g_log_default_handler () at /usr/lib/libglib-2.0.so.0
#5  0x0000560cbe6558cc in  ()
#6  0x00007f6259f038be in g_logv () at /usr/lib/libglib-2.0.so.0
#7  0x00007f6259f03ac0 in g_log () at /usr/lib/libglib-2.0.so.0
#8  0x00007f62596f28cb in  () at /usr/lib/libgjs.so.0
#9  0x00007f62596a6f38 in  () at /usr/lib/libgjs.so.0
#10 0x00007f62596ee3f5 in  () at /usr/lib/libgjs.so.0
#11 0x00007f6256ebfead in  () at /usr/lib/libmozjs-68.so
#12 0x00007f6256ee5e42 in  () at /usr/lib/libmozjs-68.so
#13 0x00007f6256ee6263 in  () at /usr/lib/libmozjs-68.so
#14 0x00007f6256ec8a62 in  () at /usr/lib/libmozjs-68.so
#15 0x00007f6256ecc81d in  () at /usr/lib/libmozjs-68.so
#16 0x00007f6256ecceda in  () at /usr/lib/libmozjs-68.so
#17 0x00007f6256d794f7 in JS_GC(JSContext*, JS::GCReason) () at /usr/lib/libmozjs-68.so
#18 0x00007f62596d5670 in  () at /usr/lib/libgjs.so.0
#19 0x00007f6259f06684 in  () at /usr/lib/libglib-2.0.so.0
#20 0x00007f6259f0b88f in g_main_context_dispatch () at /usr/lib/libglib-2.0.so.0
#21 0x00007f6259f0d831 in  () at /usr/lib/libglib-2.0.so.0
#22 0x00007f6259f0e843 in g_main_loop_run () at /usr/lib/libglib-2.0.so.0
#23 0x00007f6259406349 in meta_run () at /usr/lib/libmutter-6.so.0
#24 0x0000560cbe65541e in  ()
#25 0x00007f625a463023 in __libc_start_main () at /usr/lib/libc.so.6
#26 0x0000560cbe65561e in  ()
```

Finii commented:
Yes. I have a crash after prolonged click-and-hold on ``+`` or ``-`` of opacity. Not on the other settings, but they reach a limit rather quick.

terencode commented:
Hmm, is this an upstream issue?

Finii commented:
@terencode: Need more information than in the commit message (click on the commit hash: 
0b4cf38 )? Can explain further, but I though that was an obvious error in the code.
Sometimes your own old code looks so alien and you ask yourself who would write such a mess. Old means like a week old ,)

Finii commented:
Finally -=> icon size selectable by settings dialog.

The previous solution only allowed to set a maximum size, which means just the possibility to make the icons smaller; the new solution directly sets the selected icon size. This has been requested by @Spielkind, where our icons were smaller than the system icons.

If the selected size is bigger than for exampe the bar height, they will get distorted (i.e. just grow wider, not higher).

The default value is zero (0). This is a special value where the extension works as before, using the system default as size.

spielkind commented:
@3v1n0 what is missing to get this merged?

terencode commented:
@Finii nice!
While you are at it, maybe you can add a placement option using https://github.com/ubuntu/gnome-shell-extension-appindicator/blob/master/indicatorStatusIcon.js#L119 so I can choose to place the indicators to the left of the statusbar for example.

Finii commented:
Last commit fixes a requests I have seen (I think) a lot, (but maybe I'm mistaken).
There seems to be no one issue I can link to.
At least @terencode will be happy ;) Thanks for the exact line to change, that was helpful :-)

If this is too much customibility for @3v1n0: All the customization commits should be quite self contained and one can drop whatever is deemed unneeded.

terencode commented:
@Finii yes I will be very happy ^^ I actually also had the line for changing the icon size but forgot to share...

Finii commented:
Some small changes to make pulling more probable.
I would have requested these changes ;)

Finii commented:
Rebase on master. Force push.

terencode commented:
This change makes the appindicator from antimicrox (https://github.com/juliagoda/antimicroX) too big:

Before 196:
![image](https://user-images.githubusercontent.com/12665563/77648772-f8d1c380-6f68-11ea-8e5f-f3da6d545d20.png)

After:
![image](https://user-images.githubusercontent.com/12665563/77648713-e061a900-6f68-11ea-9610-f8ae3a7717ea.png)

Finii commented:
@terencode: what is your size setting? Looks like being 26 or so?

Finii commented:
What is that keyhole icon, seems untouched? Also generated by us? In that case I would check antiminox out ;-)

terencode commented:
> what is your size setting? Looks like being 46 or so? 

Not even close, it's 24 but the antimicrox icon size somehow doesn't change at all while I'm changing the icon size while the other do.

>Try setting size to zero (0) = use system default.

Doesn't improve at all and makes the icon blurry.
![image](https://user-images.githubusercontent.com/12665563/77659079-e57a2480-6f77-11ea-93bd-fc733c535742.png)

>What is that keyhole icon, seems untouched? 

It's psensors (https://gitlab.com/jeanfi/psensor)
Not sure what you mean by untouched, it behaves correctly (resizes), like the other one (cadence).

Finii commented:
Thanks for the input, Ill check tonight.

Finii commented:
Hmm, I don't see this with #196, but I see what you describe with #219.
Just to make sure, you do not mix the PRs up?
![Screenshot from 2020-03-26 17-15-09](https://user-images.githubusercontent.com/16012374/77669688-75729b00-6f85-11ea-9853-dbda67153927.png)

With #219 I would have a hunch...

terencode commented:
Oups sorry I posted in the wrong PR.... yes it is indeed regarding #219

Finii commented:
:-)

This PR is still able/ready to be pulled. (Just to clarify.)

Finii commented:
Fixed resizing for pixmap based icons.

Please note commit message (note at bottom) of f56f96d.

Finii commented:
Thanks for your confidence in me @thankjura :-)
But just because I fixed 1 and a half things does not qualify me as maintainer.

And forks are not the best solution for users. I usually hate stuff that gets forked, multiple times in the worst case, parallel developing solutions in different directions that the users would rather see united.

I think there is a demand for this feature, when you look at the recent comments on extensions.gnome.org. But it is a big(ger) PR and 3v1n0 needs to make sure it is maintainable, fits the general direction of development, does not introduce stuff that leads to many complaints etc pp.

3v1n0 reviewed (changes requested):
Uff I just noticed I commented things long time ago but never submitted them....

Sorry for, this maybe you can still handle them?
Otherwise I can.

3v1n0 reviewed (commented):
What kind of scenario should this fix?

As I've been doing some tests, but no difference with this commit

Finii commented:
I remember there was one application in the set I tested that needed that.
And then I though my commit messages are already long; but now vital information is missing :-|
Will try to find the machine I did the tests on.

Thank you for reviving this PR.

Finii commented:
> What kind of scenario should the commit [f56f96d](https://github.com/ubuntu/gnome-shell-extension-appindicator/commit/f56f96d0fc7752b42f26fc78ecfce4208bc8f358) fix?
> 
> As I've been doing some tests, but no difference with this commit

Well, after understanding my own commit message...

To reproduce the wrong behavior you can:
- clone https://github.com/juliagoda/antimicroX.git
- use master, I just checked the current HEAD 968c925b (I used 9b8fccf2 back then)
- `cd antimicroX && mkdir build && cd build && cmake .. && make`
- start the application in `build/bin`, the top bar icon appears
- play with the size setting, esp. values 0, 1, 2, 18, 22

I remember it took me long to find and understand this one. It is because antimicroX has only a pixmap icon that can be identified as overlay or normal icon. We would rather prefer the normal use.

Here how it looks on my machine. The icons left to right are: Printers, antimicroX, Remmina, nomachine, Caffeine. Note that Printers and Caffeine are not handled by this extension (appindicator).

With the commit in question for sizes 0, 1, 2, 18 and 22:
![ob](https://user-images.githubusercontent.com/16012374/93570810-de1a5d80-f993-11ea-8943-99ee40ea49a9.png)

Without the commit for the same sizes 0, 1, 2, 18 and 22. Note the wrong size and strange behavior at '1':
![0](https://user-images.githubusercontent.com/16012374/93570883-f68a7800-f993-11ea-8eff-544cd5cf3566.png)

Finii commented:
Rebase on master.

Aforementioned commit f56f96d is now df2db109.

3v1n0 commented on appIndicator.js:
```diff
@@ -679,7 +679,7 @@ class AppIndicators_IconActor extends St.Icon {
         let opacityValue = this._settings.get_int('icon-opacity');
 
-        this.opacity = opacityValue;
+        this.set_opacity(opacityValue);
```
Well, there's no much difference honestly... In gnome-shell upstream code we tend to use properties instead.

It's true that properties don't give errors if you try to call something non-existent.

So fine as this

3v1n0 commented on indicatorStatusIcon.js:
```diff
@@ -65,6 +66,9 @@ class AppIndicators_IndicatorStatusIcon extends PanelMenu.Button {
             this._display()
+
+        this._settings = Convenience.getSettings();
+        Util.connectSmart(this._settings, 'changed::tray-pos', this, '_display');
```
Use `this._display` as callback, the string versions aren't really easy to maintain

Finii replied:
Fixed. See comment on connectSmart further up.

Finii replied:
My fix did not work. Do not know how to proceed.

3v1n0 commented on indicatorStatusIcon.js:
```diff
@@ -116,7 +120,9 @@ class AppIndicators_IndicatorStatusIcon extends PanelMenu.Button {
         this._updateMenu();
 
-        Main.panel.addToStatusArea("appindicator-"+this._indicator.uniqueId, this, 1, 'right')
+        let our_id = "appindicator-"+this._indicator.uniqueId;
```
Use <code>\`appindicator-${this._indicator.uniqueId\`</code> instead

Finii replied:
Well, this (old line 119) has been introduced with commit 5289ceb, and I just kept it.
But I changed it now.

3v1n0 commented on appIndicator.js:
```diff
@@ -303,15 +297,23 @@ class AppIndicators_IconActor extends St.Icon {
 
         this._indicator     = indicator
         this._iconSize      = icon_size
+        this._iconSize0     = icon_size
```
Does this make sense?

Finii replied:
Yes, I suppose you mean:

```diff
--- a/appIndicator.js
+++ b/appIndicator.js
@@ -21,6 +21,7 @@ const Gio = imports.gi.Gio
 const GLib = imports.gi.GLib
 const GObject = imports.gi.GObject
 const Gtk = imports.gi.Gtk
+const Panel = imports.ui.panel;
 const St = imports.gi.St
 const Shell = imports.gi.Shell
 
@@ -347,7 +348,6 @@ class AppIndicators_IconActor extends St.Icon {
 
         this._indicator     = indicator
         this._iconSize      = icon_size
-        this._iconSize0     = icon_size;
         this._iconCache     = new IconCache.IconCache()
         this._cancellable   = new Gio.Cancellable();
         this._loadingIcons  = new Set();
@@ -772,6 +772,6 @@ class AppIndicators_IconActor extends St.Icon {
         if (sizeValue > 0)
             this._iconSize = sizeValue;
         else
-            this._iconSize = this._iconSize0;
+            this._iconSize = Panel.PANEL_ICON_SIZE;
     }
 });
```

This is of course possible, but then if for some unlikely reason
```
this._iconBox = new AppIndicator.IconActor(indicator, Panel.PANEL_ICON_SIZE);
```
is changed it will fail in strange ways. We can of course change that, but maybe the current solution is a bit more robust at some memory cost.

3v1n0 commented on appIndicator.js:
```diff
@@ -326,6 +326,11 @@ class AppIndicators_IconActor extends St.Icon {
         Util.connectSmart(Gtk.IconTheme.get_default(), 'changed', this, '_invalidateIcon')
 
+        Util.connectSmart(this, 'enter-event', this, () => {
+            this.opacity = 255;
```
Didn't you prefer the methods? :smile:

Finii replied:
This is already the case, changed with later commit "appIndicator: Use setter for opacity".

3v1n0 commented on appIndicator.js:
```diff
@@ -326,6 +326,11 @@ class AppIndicators_IconActor extends St.Icon {
+        Util.connectSmart(this, 'enter-event', this, () => {
+            this.opacity = 255;
+        });
+        Util.connectSmart(this, 'leave-event', this, '_setOpacity');
```
Use function without quotes

3v1n0 replied:
Actually, there's no need to use `Util.connectSmart` here as you're connecting to the same object, so just use simple arrow-functions (there's no need to disconnect from self on destruction).

3v1n0 commented on appIndicator.js:
```diff
@@ -352,8 +352,6 @@ class AppIndicators_IconActor extends St.Icon {
     // collector.
     _cacheOrCreateIconByName(iconSize, iconName, themePath, callback) {
         let {scale_factor} = St.ThemeContext.get_for_stage(global.stage);
-        let maxSize = Convenience.getSettings().get_int('icon-size');
```
Mmhmh, please don't add this in the first place, there's no point for uptream history to add and then remove a change in the same branch.

Finii replied:
Cleaned up all commits by squashing appropriate changes.
Now the exact history of how this has been developed is not visible anymore, but that is not useful anyhow.
Kept small fixes as these commits carry code explanations.

Finii commented:
Again rebase on master.
Clean up the commits, remove intermediate (later reverted) commits etc.
Now in a state to work on code issues.

Finii commented on appIndicator.js:
```diff
@@ -303,15 +297,23 @@ class AppIndicators_IconActor extends St.Icon {
+        Util.connectSmart(this._settings, 'changed::icon-brightness', this, '_invalidateIcon')
+        Util.connectSmart(this._settings, 'changed::icon-contrast', this, '_invalidateIcon')
+        Util.connectSmart(this._settings, 'changed::icon-size', this, '_invalidateIcon');
+
```
Changed.
Albeit all other connectSmart() calls use this quotes way and I wanted to keep the existing style.

Finii replied:
My fix did not work. Do not know how to proceed.

Finii commented:
Added some (4) commits that fix stuff you mentioned in the review. I split the commits up in a way so that it is easy to `fixup` them in a rebase, to make the changes invisible and the commits more clear.

I'm not really sure about the string function stuff, because the prevalent style is the one I used.

I can squash (fixup) the commits in and force push, if you say this is good.

The other items on the list will be touched tomorrow.

Finii commented:
New day, new energy.
Unfortunately yesterdays changes did not work. When not using the string based smartConnect, the actual functions seem to get not connected. Maybe I use that wrongly? I don't know. Have to remove the commits.

Furthermore after rebasing on master we have a regression with the wrong pixmap icon size after restart.

Finii commented on appIndicator.js:
```diff
@@ -648,4 +660,43 @@ class AppIndicators_IconActor extends St.Icon {
+            this._iconSize = sizeValue;
+        else
+            this._iconSize = this._iconSize0;
+    }
```
The code above is in `_setIconSize()`.

If it would call `_invalidateIcon()`, that calls `_updateIcon()` which in turn calls `_setIconSize()`, and we have an infinite loop ;)

All changes on the icons are triggered by a call to `_invalidateIcon()`, there is never a direct call to `_setIconSize()` (and the others), so there is no need to invalidate it afterwards. It is already invalidated if this function (or any `_set*()`) is called.

Finii commented on appIndicator.js:
```diff
@@ -648,4 +660,43 @@ class AppIndicators_IconActor extends St.Icon {
+            sat_effect = new Clutter.DesaturateEffect({});
+            this.add_effect_with_name('desaturate', sat_effect);
+        }
+        sat_effect.set_factor(desaturationValue);
```
done with commit "appIndicator: Add effect only if it is needed"

Finii commented on appIndicator.js:
```diff
@@ -648,4 +660,43 @@ class AppIndicators_IconActor extends St.Icon {
+            this.add_effect_with_name('brightness-contrast', bright_effect);
+        }
+        bright_effect.set_brightness(brightnessValue);
+        bright_effect.set_contrast(contrastValue);
```
done with commit "appIndicator: Add effect only if it is needed"

Finii commented:
There is still an issue with pixmap icons (i.e. antimicroX), the commit we discussed above seems to be not sufficient anymore, after restart or unlocking the icons seem to be created without going thru invalidate. I will examine Thurday the earliest.

As for the other comments you left in the review, I am not sure I am able to do anything there. Will comment on them on Thursday.

Thank you, Fini

terencode commented:
Hey @Finii do you think you can rebase this?

Finii commented:
Of course @terencode . But this is untested, because I have not the environment handy.

Finii commented:
Oh my gosh ... ESLint :grimacing:

terencode commented:
> Of course @terencode . But this is untested, because I have not the environment handy.

Thank ya! Tried it and got:
```
 JS ERROR: ReferenceError: icon_size is not defined
                                                    _init@/home/terence/.local/share/gnome-shell/extensions/appindicatorsupport@rgcjonas.gmail.com/appIndicator.js:387:9
```

So then I changed `this._iconSize0     = icon_size;` to ` this._iconSize0     = iconSize;` but then I got this: 
```
 JS ERROR: TypeError: Unsupported function
                                                    connectSmart4A@/home/terence/.local/share/gnome-shell/extensions/appindicatorsupport@rgcjonas.gmail.com/util.js:236:15
                                                    connectSmart@/home/terence/.local/share/gnome-shell/extensions/appindicatorsupport@rgcjonas.gmail.com/util.js:268:16
                                                    _init@/home/terence/.local/share/gnome-shell/extensions/appindicatorsupport@rgcjonas.gmail.com/appIndicator.js:397:14
                                                    _init@/home/terence/.local/share/gnome-shell/extensions/appindicatorsupport@rgcjonas.gmail.com/indicatorStatusIcon.js:45:25
```

and here I don't know how to fix it.

> Oh my gosh ... ESLint 

I'm hoping there is a fix to fix common formatting automatically? :grimacing:

3v1n0 commented:
> Oh my gosh ... ESLint grimacing

Eh, I know :joy:, but that's the only viable way to have some code that is maintainable in the long run and with some static checks :)

Finii commented:
Hmm.
I believe it works now.
Still some style issues, mostly `Lang.bind` and unused vars.
Need some sleep (it's 3:17 am here).
Ping me if more help needed. :sleeping:

Finii commented:
Finally :tada: all ESLint stuff squelshed and it is even working on my machine (Ubuntu 20.10).

Maybe it gets pulled somewhen ;-)

terencode commented:
Tested and working for me too, thanks!

Not sure if directly related to the extension but when I opened the settings for the first time I got `gjs[354945]: g_task_return_error: assertion '!task->ever_returned' failed`

Finii commented:
Force push: rebase on master

3v1n0 commented:
Thanks for keeping this maintained and the original work, I've finally got time to look deeply at it and merge.

I've done some changes though, and squashed it mostly to one single commit as the various commits in the middle were not relevant for master branch.

A part the refactoring commits I've applied to master too, the changes I did to your initial proposal are these:

<details>

```diff
diff --git a/.gitignore b/.gitignore
index 378eac2..bde8064 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1 +1,3 @@
 build
+schemas/gschemas.compiled
+*.mo
diff --git a/appIndicator.js b/appIndicator.js
index 1b6197b..fd827f4 100644
--- a/appIndicator.js
+++ b/appIndicator.js
@@ -385,7 +385,6 @@ class AppIndicatorsIconActor extends St.Icon {
 
         this._indicator     = indicator;
         this._iconSize      = iconSize;
-        this._iconSize0     = iconSize;
         this._iconCache     = new IconCache.IconCache();
         this._cancellable   = new Gio.Cancellable();
         this._loadingIcons  = new Set();
@@ -413,10 +412,14 @@ class AppIndicatorsIconActor extends St.Icon {
 
         Util.connectSmart(Gtk.IconTheme.get_default(), 'changed', this, this._invalidateIcon);
 
-        Util.connectSmart(this, 'enter-event', this, () => {
-            this.set_opacity(255);
+        this.connect('enter-event', () => {
+            this.opacity = 255;
+            return Clutter.EVENT_PROPAGATE;
+        });
+        this.connect('leave-event', () => {
+            this._setOpacity();
+            return Clutter.EVENT_PROPAGATE;
         });
-        Util.connectSmart(this, 'leave-event', this, this._setOpacity);
 
         if (indicator.isReady)
             this._invalidateIcon();
@@ -474,7 +477,7 @@ class AppIndicatorsIconActor extends St.Icon {
             const inputStream = await file.read_async(GLib.PRIORITY_DEFAULT, this._cancellable);
             const pixbuf = GdkPixbuf.Pixbuf.new_from_stream_at_scale_async(inputStream,
                 height, width, true, this._cancellable);
-            this.iconSize = width > 0 ? width : this._iconSize;
+            this.icon_size = width > 0 ? width : this._iconSize;
             return pixbuf;
         } catch (e) {
             if (!e.matches(Gio.IOErrorEnum, Gio.IOErrorEnum.CANCELLED))
@@ -520,7 +523,7 @@ class AppIndicatorsIconActor extends St.Icon {
                 /* Hello indicator-multiload! */
                 return this._createIconByPath(path, width, -1);
             } else {
-                this.iconSize = this._iconSize;
+                this.icon_size = this._iconSize;
                 return new Gio.FileIcon({
                     file: Gio.File.new_for_path(path),
                 });
@@ -666,13 +669,15 @@ class AppIndicatorsIconActor extends St.Icon {
     // one (as in some cases it may be equal, but not the same object).
     // So when it's not need anymore we make sure to check the .inUse property
     // and set it to false so that it can be picked up by the garbage collector.
-    _setGicon(iconType, gicon) {
+    _setGicon(iconType, gicon, iconSize) {
         if (iconType !== SNIconType.OVERLAY) {
             if (gicon) {
                 this.gicon = new Gio.EmblemedIcon({ gicon });
 
                 if (!(gicon instanceof GdkPixbuf.Pixbuf))
                     gicon.inUse = this.gicon.get_icon() === gicon;
+
+                this.set_icon_size(iconSize);
             } else {
                 this.gicon = null;
                 Util.Logger.critical(`unable to update icon for ${this._indicator.id}`);
@@ -721,8 +726,7 @@ class AppIndicatorsIconActor extends St.Icon {
                 gicon = await this._createIconFromPixmap(iconSize, pixmap, iconType);
             }
 
-            this._setGicon(iconType, gicon);
-            this.set_iconSize(iconSize);
+            this._setGicon(iconType, gicon, iconSize);
         } catch (e) {
             /* We handle the error messages already */
             if (!e.matches(Gio.IOErrorEnum, Gio.IOErrorEnum.CANCELLED) &&
@@ -773,56 +777,56 @@ class AppIndicatorsIconActor extends St.Icon {
         this._iconCache.clear();
         this._cancelLoading();
 
-        this._updateOverlayIcon();
         this._updateIcon();
+        this._updateOverlayIcon();
     }
 
-
     _setOpacity() {
-        let opacityValue = this._settings.get_int('icon-opacity');
-
-        this.set_opacity(opacityValue);
+        this.opacity = this._settings.get_int('icon-opacity');
     }
 
     _setSaturation() {
-        let desaturationValue = this._settings.get_double('icon-saturation');
-
-        let satEffect = this.get_effect('desaturate');
+        const desaturationValue = this._settings.get_double('icon-saturation');
+        let desaturateEffect = this.get_effect('desaturate');
 
         if (desaturationValue > 0) {
-            if (!satEffect) {
-                satEffect = new Clutter.DesaturateEffect({});
-                this.add_effect_with_name('desaturate', satEffect);
+            if (!desaturateEffect) {
+                desaturateEffect = new Clutter.DesaturateEffect();
+                this.add_effect_with_name('desaturate', desaturateEffect);
             }
-            satEffect.set_factor(desaturationValue);
-        } else if (satEffect) {
-            this.remove_effect(satEffect);
+            desaturateEffect.set_factor(desaturationValue);
+        } else if (desaturateEffect) {
+            this.remove_effect(desaturateEffect);
         }
     }
 
     _setBrightnessContrast() {
-        let brightnessValue = this._settings.get_double('icon-brightness');
-        let contrastValue = this._settings.get_double('icon-contrast');
-
-        let brightEffect = this.get_effect('brightness-contrast');
+        const brightnessValue = this._settings.get_double('icon-brightness');
+        const contrastValue = this._settings.get_double('icon-contrast');
+        let brightnessContrastEffect = this.get_effect('brightness-contrast');
 
         if (brightnessValue !== 0 | contrastValue !== 0) {
-            if (!brightEffect) {
-                brightEffect = new Clutter.BrightnessContrastEffect({});
-                this.add_effect_with_name('brightness-contrast', brightEffect);
+            if (!brightnessContrastEffect) {
+                brightnessContrastEffect = new Clutter.BrightnessContrastEffect();
+                this.add_effect_with_name('brightness-contrast', brightnessContrastEffect);
             }
-            brightEffect.set_brightness(brightnessValue);
-            brightEffect.set_contrast(contrastValue);
-        } else if (brightEffect) {
-            this.remove_effect(brightEffect);
+            brightnessContrastEffect.set_brightness(brightnessValue);
+            brightnessContrastEffect.set_contrast(contrastValue);
+        } else if (brightnessContrastEffect) {
+            this.remove_effect(brightnessContrastEffect);
         }
     }
 
     _setIconSize() {
         let sizeValue = this._settings.get_int('icon-size');
-        if (sizeValue > 0)
+        if (sizeValue > 0) {
+            if (!this._defaultIconSize)
+                this._defaultIconSize = this._iconSize;
+
             this._iconSize = sizeValue;
-        else
-            this._iconSize = this._iconSize0;
+        } else if (this._defaultIconSize) {
+            this._iconSize = this._defaultIconSize;
+            delete this._defaultIconSize;
+        }
     }
 });
diff --git a/indicatorStatusIcon.js b/indicatorStatusIcon.js
index ed4eb68..65d86d8 100644
--- a/indicatorStatusIcon.js
+++ b/indicatorStatusIcon.js
@@ -69,7 +69,6 @@ class AppIndicatorsIndicatorStatusIcon extends PanelMenu.Button {
 
         if (this._indicator.isReady)
             this._display();
-
     }
 
     _updateLabel() {
diff --git a/locale/de/LC_MESSAGES/AppIndicator-Support.po b/locale/de/LC_MESSAGES/AppIndicatorExtension.po
similarity index 100%
rename from locale/de/LC_MESSAGES/AppIndicator-Support.po
rename to locale/de/LC_MESSAGES/AppIndicatorExtension.po
diff --git a/locale/fr/LC_MESSAGES/AppIndicator-Support.po b/locale/fr/LC_MESSAGES/AppIndicatorExtension.po
similarity index 100%
rename from locale/fr/LC_MESSAGES/AppIndicator-Support.po
rename to locale/fr/LC_MESSAGES/AppIndicatorExtension.po
diff --git a/locale/hu/LC_MESSAGES/AppIndicator-Support.po b/locale/hu/LC_MESSAGES/AppIndicatorExtension.po
similarity index 100%
rename from locale/hu/LC_MESSAGES/AppIndicator-Support.po
rename to locale/hu/LC_MESSAGES/AppIndicatorExtension.po
diff --git a/locale/it/LC_MESSAGES/AppIndicator-Support.po b/locale/it/LC_MESSAGES/AppIndicatorExtension.po
similarity index 100%
rename from locale/it/LC_MESSAGES/AppIndicator-Support.po
rename to locale/it/LC_MESSAGES/AppIndicatorExtension.po
diff --git a/locale/ja/LC_MESSAGES/AppIndicator-Support.po b/locale/ja/LC_MESSAGES/AppIndicatorExtension.po
similarity index 100%
rename from locale/ja/LC_MESSAGES/AppIndicator-Support.po
rename to locale/ja/LC_MESSAGES/AppIndicatorExtension.po
diff --git a/locale/nl/LC_MESSAGES/AppIndicator-Support.po b/locale/nl/LC_MESSAGES/AppIndicatorExtension.po
similarity index 100%
rename from locale/nl/LC_MESSAGES/AppIndicator-Support.po
rename to locale/nl/LC_MESSAGES/AppIndicatorExtension.po
diff --git a/locale/pt_BR/LC_MESSAGES/AppIndicator-Support.po b/locale/pt_BR/LC_MESSAGES/AppIndicatorExtension.po
similarity index 100%
rename from locale/pt_BR/LC_MESSAGES/AppIndicator-Support.po
rename to locale/pt_BR/LC_MESSAGES/AppIndicatorExtension.po
diff --git a/locale/ru/LC_MESSAGES/AppIndicator-Support.po b/locale/ru/LC_MESSAGES/AppIndicatorExtension.po
similarity index 100%
rename from locale/ru/LC_MESSAGES/AppIndicator-Support.po
rename to locale/ru/LC_MESSAGES/AppIndicatorExtension.po
diff --git a/locale/sr/LC_MESSAGES/AppIndicator-Support.po b/locale/sr/LC_MESSAGES/AppIndicatorExtension.po
similarity index 100%
rename from locale/sr/LC_MESSAGES/AppIndicator-Support.po
rename to locale/sr/LC_MESSAGES/AppIndicatorExtension.po
diff --git a/locale/sr@latin/LC_MESSAGES/AppIndicator-Support.po b/locale/sr@latin/LC_MESSAGES/AppIndicatorExtension.po
similarity index 100%
rename from locale/sr@latin/LC_MESSAGES/AppIndicator-Support.po
rename to locale/sr@latin/LC_MESSAGES/AppIndicatorExtension.po
diff --git a/locale/tr/LC_MESSAGES/AppIndicator-Support.po b/locale/tr/LC_MESSAGES/AppIndicatorExtension.po
similarity index 100%
rename from locale/tr/LC_MESSAGES/AppIndicator-Support.po
rename to locale/tr/LC_MESSAGES/AppIndicatorExtension.po
diff --git a/locale/zh_CN/LC_MESSAGES/AppIndicator-Support.po b/locale/zh_CN/LC_MESSAGES/AppIndicatorExtension.po
similarity index 100%
rename from locale/zh_CN/LC_MESSAGES/AppIndicator-Support.po
rename to locale/zh_CN/LC_MESSAGES/AppIndicatorExtension.po
diff --git a/metadata.json b/metadata.json
index dc54bbb..5cacfa0 100644
--- a/metadata.json
+++ b/metadata.json
@@ -6,8 +6,8 @@
         "3.38",
         "40"
     ],
-    "gettext-domain": "AppIndicator-Support",
-    "settings-schema": "org.gnome.shell.extensions.appindicatorsupport",
+    "gettext-domain": "AppIndicatorExtension",
+    "settings-schema": "org.gnome.shell.extensions.appindicator",
     "uuid": "appindicatorsupport@rgcjonas.gmail.com",
     "name": "KStatusNotifierItem/AppIndicator Support",
     "description": "Adds KStatusNotifierItem support to the Shell",
diff --git a/prefs.js b/prefs.js
index a4f91dc..ae5fec3 100644
--- a/prefs.js
+++ b/prefs.js
@@ -129,47 +129,6 @@ const AppIndicator = new GObject.Class({
         this._settings.bind('tray-pos', widget, 'active-id', Gio.SettingsBindFlags.DEFAULT);
         this.attach(label, 0, 7, 1, 1);
         this.attach(widget, 1, 7, 1, 1);
-
-        // comment unused stuff out
-        //
-        // // Icon tray spacing
-        // label = new Gtk.Label({
-        //     label: _('Spacing between icons (min: 0, max: 20)'),
-        //     hexpand: true,
-        //     halign: Gtk.Align.START,
-        // });
-        // widget = new Gtk.SpinButton({ halign: Gtk.Align.END });
-        // widget.set_sensitive(true);
-        // widget.set_range(0, 20);
-        // widget.set_value(this._settings.get_int('icon-spacing'));
-        // widget.set_increments(1, 2);
-        // widget.connect('value-changed', w => {
-        //     this._settings.set_int('icon-spacing', w.get_value_as_int());
-        // });
-        // this.attach(label, 0, 6, 1, 1);
-        // this.attach(widget, 1, 6, 1, 1);
-        //
-        // // Tray order in panel
-        // label = new Gtk.Label({
-        //     label: _('Tray offset'),
-        //     hexpand: true,
-        //     halign: Gtk.Align.START,
-        // });
-        // widget = new Gtk.SpinButton({ halign: Gtk.Align.END });
-        // widget.set_sensitive(true);
-        // widget.set_range(0, 20);
-        // widget.set_value(this._settings.get_int('tray-order'));
-        // widget.set_increments(1, 2);
-        // widget.connect('value-changed', w => {
-        //     this._settings.set_int('tray-order', w.get_value_as_int());
-        // });
-        // this.attach(label, 0, 8, 1, 1);
-        // this.attach(widget, 1, 8, 1, 1);
-        //
-        // }
-        // comment unused stuff out
-
-        // this._changedPermitted = true;
     },
 
 });
diff --git a/schemas/org.gnome.shell.extensions.appindicator.gschema.xml b/schemas/org.gnome.shell.extensions.appindicator.gschema.xml
index edad5d9..55144ee 100644
--- a/schemas/org.gnome.shell.extensions.appindicator.gschema.xml
+++ b/schemas/org.gnome.shell.extensions.appindicator.gschema.xml
@@ -1,7 +1,7 @@
-<schemalist gettext-domain="AppIndicator-Support">
-  <schema id="org.gnome.shell.extensions.appindicatorsupport" path="/org/gnome/shell/extensions/appindicatorsupport/">
+<schemalist gettext-domain="AppIndicatorExtension">
+  <schema id="org.gnome.shell.extensions.appindicator" path="/org/gnome/shell/extensions/appindicator/">
     <key name="icon-saturation" type="d">
-      <default>0.4</default>
+      <default>0.0</default>
       <summary>Saturation</summary>
     </key>
     <key name="icon-brightness" type="d">
@@ -13,7 +13,7 @@
       <summary>Contrast</summary>
     </key>
     <key name="icon-opacity" type="i">
-      <default>220</default>
+      <default>240</default>
       <summary>Opacity</summary>
     </key>
     <key name="icon-size" type="i">
```
</details>

Finii commented:
Thank you for merging this :+1:.

I kept the small commits to make following the change steps easier, but squashing them on merge is what I expected.

Have some comments that are so small that I hesitate to raise issues:

1: `Makefile`: `rm -rf ls -d local/*/` is probably a typo (the `ls` part) (or course I liked the double-* syntax better ;)

2: Un-desaturate on hover is an excellet addition. But for the colour icon probably we need to also deactivate the other settings like contrast and brightness? Usually you use these two to adapt the monochrome icon; on a colour icon they usually look pretty bad.

I can create PRs if you prefer.

Thank you again!

Finii commented:
Just noticed...

![image](https://user-images.githubusercontent.com/16012374/121126934-44e4ed00-c829-11eb-862b-905adfd052de.png)

One time _Kstatus_ first _AppIndicator_ second (Gnome Tweaks' extensions list), one time the other way round (Settings dialog).

Arcitec commented:
@Finii Thank you a few years later. My mom is 70 and your ability to set the tray icon size means she no longer has to put on glasses and lean forwards to see the icons. :)
